LegendaryExplorer/LegendaryExplorer/Tools/PackageEditor/Experiments/PackageEditorExperimentsO.cs

```diff
@@ -688,5 +689,381 @@ public static void Baldinator(PackageEditorWindow pew)
+            }
+
+            // Prompt for and validate origin position
+            string promptOrigin = PromptDialog.Prompt(null, "Write the origin coordinates from which to remove references to the mesh. It must be a comma separated list: X, Y, Z");
+            if (string.IsNullOrEmpty(promptOrigin))
+            {
+                ShowError("Invalid origin");
+                return;
+            }
+
+            string[] strOrigins = promptOrigin.Split(",");
+            if (strOrigins.Length != 3)
+            {
+                ShowError("Wrong number of coordinates. You must provide 3 coordinates in the form of: X, Y, Z");
+                return;
+            }
+            List<float> origins = new();
+            foreach (string strOrigin in strOrigins)
+            {
+                float origin;
+                if (!float.TryParse(strOrigin, out origin))
+                {
+                    ShowError($"Distance {strOrigin} is not a valid decimal");
+                    return;
+                }
+                origins.Add(origin);
+            }
+
+            // Prompt for and validate distance
+            string promptDist = PromptDialog.Prompt(null, "Write the distance from each origin coordinate in which to remove references. It must be a comma separated list: distX, distY, distZ");
+            if (string.IsNullOrEmpty(promptDist))
+            {
+                ShowError("Invalid distance");
+                return;
+            }
+
+            string[] strDists = promptDist.Split(",");
+            if (strDists.Length != 3)
+            {
+                ShowError("Wrong number of distances. You must provide 3 distances, in the form of: distX, distY, distZ");
+                return;
+            }
+            List<float> dists = new();
+            foreach (string strDist in strDists)
+            {
+                float dist;
+                if (!float.TryParse(strDist, out dist))
+                {
+                    ShowError($"Distance {strDist} is not a valid decimal");
+                    return;
+                }
+                if (dist < 0)
+                {
+                    ShowError($"Distance {strDist} must be a positive value.");
+                    return;
+                }
+                dists.Add(dist);
+            }
+
+            ExportEntry mesh = (ExportEntry)pew.SelectedItem.Entry;
+            // Get a list of SMC/SMAs referencing the selected mesh
+            List<IEntry> references = mesh.GetEntriesThatReferenceThisOne()
+                .Where(kvp => kvp.Key.ClassName is "SkeletalMeshComponent" or "StaticMeshComponent")
+                .Select(kvp => kvp.Key).ToList();
+
+            if (references.Count == 0)
+            {
+                ShowError("Found no SMCs referencing the selected mesh");
+                return;
+            }
+
+            List<string> removedReferences = new();
+            foreach (ExportEntry reference in references)
```

Comment:
Something to watch out for with `foreach` loops is that it will (unfortunately) let you specify an iteration variable of a different type than is contained in the collection, and will insert an invisible cast. `references` contains `IEntry`s, which can be `ExportEntry`s or `ImportEntry`s. So if there was an `ImportEntry` in `references`, this would fail at runtime with an `InvalidCastException`. Now it just so happens that `GetEntriesThatReferenceThisOne` only returns `ExportEntry`s, despite its signature suggesting otherwise, so in this case it will always work! But in general, you should not do this.

So using this as an example, here's a couple of better ways to deal with a mismatch between the type in a collection, and the type you want.

1. Filtering using OfType
```csharp
foreach (ExportEntry reference in references.OfType<ExportEntry>())
```
This will get only the `ExportEntry`s in the collection, and ignore anything else. This is good when you don't want to do anything with other types, and don't care if they exist.

2. Type checking in the loop
```csharp
foreach (IEntry referenceEntry in references)
{
    if (referenceEntry is ExportEntry referenceExport)
    {
        //do stuff with referenceExport
    }
    else
    {
       //it's not an ExportEntry! throw an exception? do something else? you decide!
    }
}
```

Reply:
Thanks for explaining this. For some reason I was assuming that all IEntries are ExportEntries. I didn't know about OfType, it's pretty convenient.

Reply:
If you're going to be iterating over a loop, and you're going to modify it (such as `foreach(var exp in pcc.Exports)`, and adding/cloning exports), you will want to make sure you do a `.ToList()` on the end, which will get all results first and then process the loop. It's not always necessary, but you will find a fun edge case where it throws an error for modifying an enumeration. Calling .ToList() pulls all results before processing the loop, which avoids that issue.